### PR TITLE
Mimer: a ~~drop-in~~ replacement for Agsy

### LINKED ISSUE #1789: Agsy generates ill-typed terms referring to shadowed variables
The problem is that agsy does not notice that a variable it wants to use
has been shadowed and that the typechecker will therefore reject the 
generated term as ill-typed.

```
argh : (X : Set) → X → Set
argh = λ x x → ? -- Here C-c C-a yields "x"
```

Credits to @pigworker for highlighting that the machine being able to
generate valid names for shadowed variables is crucial for a smooth
interactive programming experience.

We can either patch agsy so that it refuses to produce a term in such
situations or add a way to refer to shadowed variables (e.g. "x^n" would
mean "the nth x in scope"). The latter solution would also yield a fix for
a similar problem when solving unification problems automatically:

```
argh₂ : (X : Set) (f : (A : Set) → A → Set) →
        (Y : Set) (X : X) → f ? X -- Here C-c C-s yields "X"
argh₂ = {!!} 
```

### LINKED ISSUE #2853: Auto does not work well with record types
My code here (removed irrelevant part):

```agda
bin-op : ∀ {ℓ} (A : Set ℓ) → Set ℓ
bin-op A = A → A → A

private
  record S3 {ℓ} (A : Set ℓ) : Set ℓ where
    infixl 5 _×_
    field
      x y {e} : A
      _×_ : bin-op A
      assocₗ : ∀ a b c → a × b × c ≡ a × (b × c)      

    assocᵣ : ∀ a b c → a × (b × c) ≡ a × b × c
    assocᵣ a b c rewrite assocₗ a b c = {!!}
```

I typed <kbd>C-c C-a</kbd> at the hole. Nothing happens.

Even Agda has correctly displayed the type of the hole (`?0 : a × (b × c) ≡ a × (b × c)`).

It should obviously been filled with a `refl` but it says `no solution found`. Why? Is this a bug?

### LINKED ISSUE #3674: Irrefutable pattern in Auto/CaseSplit
Don't have time to investigate right now but running auto in the hole yields

> src/full/Agda/Auto/CaseSplit.hs:468:15-41: Irrefutable pattern failed for pattern Just j

```agda
data Empty : Set where

¬ : Set → Set
¬ A = A → Empty

data Bool : Set where
  true false : Bool

record Sigma (A : Set) (B : A → Set) : Set where
  constructor _,_
  field proj₁ : A
        proj₂ : B proj₁

data _≡_ {A : Set} (a : A) : A → Set where
  refl : a ≡ a

not : (b : Bool) → Sigma Bool (λ b' → ¬ (b ≡ b'))
not b = {!-c !}
```

### LINKED ISSUE #6101: Agsy gives up when no HIT is present
MWE:
```agda
{-# OPTIONS --with-K #-}  -- will occur with or without this, also occurs with --without-K
module MWE where
open import Agda.Builtin.Equality using (_≡_)
open import Data.Nat using (ℕ; _≟_)
open import Relation.Nullary using (yes; no)

F : ℕ → ℕ
F p with p ≟ p
... | yes _ = {!   !}
... | no _ = {!   !}

swap-inj : ∀ {p} → F p ≡ F p → ℕ
swap-inj eq = {!   !}
```

When the cursor is in the last hole and <kbd>C-c C-a</kbd> is pressed, agda complains "Not implemented: The Agda synthesizer (Agsy) does not support HITs yet". If `--without-K` is enabled, then the same code works with `Bool`.

Agda version 2.6.3, using VSCode on OS X.

### LINKED ISSUE #6361: Agsy ignores --postfix-projections
```agda
{-# OPTIONS --postfix-projections #-}
module Test26 where

open import Agda.Builtin.Nat

record Test : Set where
  field
    untest : Nat

open Test

foo : Test → Nat
foo x = {!   !}
```

Agsy generates `untest x` rather than `x .untest`

### PR DESCRIPTION
**Pull request by UlfNorell**
Master's thesis report can be found here: https://odr.chalmers.se/items/aae6d187-8735-457c-89af-51117409bd51

Advantages
  - More maintainable code
  - Reusing Agda internals
  - Doesn't give up if it sees a copattern or cubical things

Limitations
  - No case splitting (`-c`)
  - No disproving (`-d`)
  - No refinement (`-r`)

Fixes #1789, fixes #2853, closes #3674 (by not having case-splitting), fixes #6101, and fixes #6361.

### CODE REVIEW

**Comment by ice1000** [comment]
image

I can't seem to open the link though...

**Comment by gallais** [comment]
Does the direct download link work for you?
https://odr.chalmers.se/server/api/core/bitstreams/0b00c9d6-eba3-45f7-a466-99c13166ffa0/content

**Comment by ice1000** [comment]
> Does the direct download link work for you? https://odr.chalmers.se/server/api/core/bitstreams/0b00c9d6-eba3-45f7-a466-99c13166ffa0/content

Yes, it works, thanks!

I am celebrating this work! It is soo nice to have a modern replacement for Agsy!

**Comment by wilbowma** [comment]
FWIW, I gave this branch a shot on the following file (from #6101) and got a Panic:
```
/private/tmp/mimer.agda:32,7-11
Panic: Unbound name:
Data.Nat.Base.<′-base[0,2,4,290]@ModuleNameHash
12988165511291312787
when checking that the expression ? has type
(oNat ∷ []) ⊢ var 0 :: oNat
```

```
open import Data.Nat

-- Represent vars as de Bruijn indexes
Var = ℕ

data Type : Set where
  -- object-language Nat
  oNat : Type

data Term : Set where
  var : Var -> Term

open import Data.List

Env = List Type

well-bound : Env -> Var -> Set
well-bound Γ n = (n < (length Γ))

open import Data.Fin hiding (_<_; _+_)

env-lookup : (Γ : Env) -> {n : Var} -> (well-bound Γ n) -> Type
env-lookup Γ p = lookup Γ (fromℕ< p)

data _⊢_::_ : Env -> Term -> Type -> Set where
  Rule-Var : ∀ {Γ n} ->
    (x : (well-bound Γ n)) ->
    ------------------------
    Γ ⊢ (var n) :: (env-lookup Γ x)

foo : (oNat ∷ []) ⊢ var 0 :: oNat
foo = {!!}
```

File: src/full/Agda/Mimer/Mimer.hs [hunk 1]
```diff
@@ -0,0 +1,1806 @@
...
+  where
+
+import Control.DeepSeq (force, NFData(..))
+import Control.Monad ((>=>), (=<<), unless, foldM, when, zipWithM, filterM)
+import Control.Monad.Except (catchError)
+import Control.Monad.Error.Class (MonadError)
+import Control.Monad.Fail (MonadFail)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Reader (ReaderT(..), runReaderT, asks, ask)
+import Data.Function (on)
```
Comment (andreasabel) [inline src/full/Agda/Mimer/Mimer.hs:14]: Good discipline, using explicit imports!

**Comment by UlfNorell** [comment]
I'm feeling (reasonably) ready to merge this now. There will be some issues I'm sure, but the easiest way to find them is to have people start using it. Any last-minute thoughts @andreasabel, @jespercockx, @gallais, @plt-amy?

**Comment by jespercockx** [comment]
I am all in favor of merging this!

**Review by jespercockx — APPROVED** [review]

**Review by plt-amy — APPROVED** [review]
I'm going to miss the name Agsy, but it's very nice to see a better integrated synthesizer 😄

**Review by andreasabel — COMMENTED** [review]
Are there any mimer tests with module parameters?   We kinda know by now that they are the source of many bugs...

Why was the Auto-Modules test deleted?  These tests look elementary (no induction, no lemmata, just DTLC).

**Pull request merged**